torchtune/utils/memory.py

```diff
@@ -6,29 +6,131 @@
 
 import gc
 
-from typing import Any, Dict, Optional, Set
+from collections import defaultdict
+
+from typing import Any, Dict
 
 import torch
 
 from torch import nn
 from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import (
-    apply_activation_checkpointing,
+    checkpoint_wrapper as ptd_checkpoint_wrapper,
+    CheckpointImpl,
 )
-from torch.distributed.fsdp.wrap import ModuleWrapPolicy
+
+from torch.utils.checkpoint import _pt2_selective_checkpoint_context_fn_gen, checkpoint
+
+# for selective AC
+no_recompute_list = {
+    torch.ops.aten.mm.default,
+    torch.ops.aten._scaled_dot_product_efficient_attention.default,
+    torch.ops.aten._scaled_dot_product_flash_attention.default,
+    torch.ops._c10d_functional.reduce_scatter_tensor.default,
+}
+
+# Uses PTD FSDP AC wrapper
+# currently selective per op and per layer checkpointing are supported
+def checkpoint_wrapper(module, ac_mode, ac_style):
+    if ac_mode == "selective" and ac_style == "op":
+
+        def _get_custom_policy(meta):
+            def _custom_policy(mode, func, *args, **kwargs):
+                mm_count_key = f"{mode}_mm_count"
+                if func == torch.ops.aten.mm.default:
+                    meta[mm_count_key] += 1
+                # Saves output of all compute ops, except every second mm
+
+                return func in no_recompute_list and not (
+                    func == torch.ops.aten.mm.default and meta[mm_count_key] % 2 == 0
+                )
```

Comment:
This block could use more comments for noobs like me :)

Reply:
np, will add!

Reply:
actually we are removing selective op ac so will resolve this.